### PR TITLE
Opengl support

### PR DESCRIPTION
**Pull request by wojciechloboda**
adding basic minimal ui and visualization for tracing using opengl.

### CODE REVIEW

File: src/LocalRenderer/Window.cpp
Comment (bartektempka) [inline src/LocalRenderer/Window.cpp:1]: Unknown error when initializing GLEW.
[Issue](https://github.com/NVlabs/nvdiffrast/issues/18) "Mea culpa, it must be that i have compiled glew without SYSTEM-linux-egl"
quick fix:
```c++
-    if (result != GLEW_OK)
+    if (result != GLEW_OK && result != GLEW_ERROR_NO_GLX_DISPLAY){
           fprintf(stderr, "Error: %s\n", glewGetErrorString(glew_status));
           throw std::runtime_error("Failed to init GLEW");
 }
```

**Review by bartektempka — CHANGES_REQUESTED** [review]

File: src/LocalRenderer/Window.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        throw std::runtime_error("Failed to create GLFW window!");
+    }
+
+    glfwMakeContextCurrent(window_.get());
+    glfwSetWindowUserPointer(window_.get(), this);
+    glfwSetScrollCallback(window_.get(), scrollCallback);
+
+    GLenum glew_status = glewInit();
+
+    if (glew_status != GLEW_OK) {
```
Comment (pklatka) [inline src/LocalRenderer/Window.cpp:33]: Just to save time with fixing the WSL issue that @bartektempka has mentioned, let's just check if a NO_GLX_DISPLAY error occurred.
```suggestion
    // GLEW_ERROR_NO_GLX_DISPLAY occurs when running the program on a WSL system.
    // This error is not critical and can be ignored.
    if (glew_status != GLEW_OK && glew_status != GLEW_ERROR_NO_GLX_DISPLAY) {
```

File: src/LocalRenderer/Utils.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+#define CheckedGLCall(x) do { printOpenGLErrors(">>BEFORE<< "#x, __FILE__, __LINE__); (x); printOpenGLErrors(#x, __FILE__, __LINE__); } while (0)
+#define CheckedGLResult(x) (x); printOpenGLErrors(#x, __FILE__, __LINE__);
+#define CheckExistingErrors(x) printOpenGLErrors(">>BEFORE<< "#x, __FILE__, __LINE__);
+
+void printOpenGLErrors(char const * const function, char const * const file, int const line) {
+	bool succeeded = true;
+	GLenum error = glGetError();
+	if (error != GL_NO_ERROR) {
+		char const *errorString = (char const *) gluErrorString(error);
```
Comment (pklatka) [inline src/LocalRenderer/Utils.h:16]:
```suggestion
	char const *errorString = (char const *) gluErrorString(error);
```

File: src/LocalRenderer/Utils.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+void printShaderInfoLog(GLint const shader) {
+	int infoLogLength = 0;
+	int charsWritten = 0;
+	glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &infoLogLength);
+	if (infoLogLength > 0) {
+		GLchar * infoLog = new GLchar[infoLogLength];
+		glGetShaderInfoLog(shader, infoLogLength, &charsWritten, infoLog);
+		std::cout << "Shader Info Log:" << std::endl << infoLog << std::endl;
```
Comment (pklatka) [inline src/LocalRenderer/Utils.h:32]: Do we need to flush the buffer here (I'm talking about std::endl) - when this function is needed?

File: src/LocalRenderer/Renderer.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <memory>
+#include <ctime>
+#include <vector>
+#include "Window.h"
+#include <GL/glew.h>
+#include <GLFW/glfw3.h>
+#include "cuda_runtime.h"
+#include <curand_kernel.h>
```
[the remark this answers is not in the document]
Comment (wojciechloboda) [reply src/LocalRenderer/Renderer.h:10]: It's not needed I tested something and forgot about it, I'll remove it.

File: src/LocalRenderer/Utils.h (hunk 3 above)
Comment (wojciechloboda) [reply src/LocalRenderer/Utils.h:32]: This is needed for debugging to check if shaders compiled properly etc, In the first version I used custom shaders so it was usefull, but then I simplified renderer and left it, we may use it in the future, and endl is there to make output more clear.

**Review by pklatka — APPROVED** [review]
LGTM!

**Pull request merged**